ownership/sync-repos.py

```diff
@@ -0,0 +1,255 @@
+)
+@click.option(
+    '--spreadsheet-csv',
+    help="Ownership spreadsheet CSV data (for local testing)"
+)
+def main(github_username, token_file, spreadsheet_csv):
+    orgs = ['edx', 'edx-solutions', 'edx-ops']
+
+    with open(token_file, 'r') as tf:
+        token = tf.readline().rstrip('\n')
+    hub = github3.login(github_username, token)
+
+    try:
+        repos_gh = list(fetch_github(hub, orgs))
+        repos_ss = list(fetch_spreadsheet(spreadsheet_csv, orgs))
+
+        run_compare(hub, orgs, repos_gh, repos_ss)
+    except KnownError as e:
+        print(e.message)
+
+
+def run_compare(hub, orgs, repos_gh, repos_ss):
+    # List of actions to recommend at end of run.
+    #
+    # This list will *grow* as the comparison progresses.
+    actions = []
+
+    # For any archived repos, we don't want them in the spreadsheet
+    # anyhow, and should recommend moving them to the edx-unsupported
+    # org.
+    archived = [r for r in repos_gh if r['archived']]
+    for r in archived:
+        actions.append({
+            'action': 'move_repo_unsupported',
+            'name': r['name'],
+            'why': ("Repo has been archived but is still in a main org "
+                    "rather than edx-unsupported")
+        })
+        repos_gh.remove(r)
+
+    names_all_gh = {r['name'] for r in repos_gh}
+    names_all_ss = {r['name'] for r in repos_ss}
+
+    # Bail out early on any duplicates, which complicate analysis
+    names_ss_dup = names_all_ss.copy() # a list of duplicated names
+    for uname in set(names_ss_dup):
+        names_ss_dup.remove(uname)
+    if names_ss_dup:
+        error_lines = [
+            ("Manual intervention required: The following repositories are "
+             "listed multiple times in the spreadsheet:")]
+        for uniq_dup_name in set(names_ss_dup):
+            row_ids = [str(r['row_id']) for r in repos_ss if r['name'] == uniq_dup_name]
+            error_lines.append("%s on rows %s" % (uniq_dup_name, ', '.join(row_ids)))
+        bail(error_lines)
+
+    # Mismatched names in both directions.
+    #
+    # These lists will *shrink* as the comparison progresses and they are
+    # converted into action items.
+    names_only_gh = names_all_gh - names_all_ss
+    names_only_ss = names_all_ss - names_all_gh
+
+    def find_row_by_name(repo_name):
+        return next(filter(lambda r: r['name'] == repo_name, repos_ss))
+    
+    def recommend_delete_row(name, why, more_data=None):
+        actions.append({
+            'action': 'delete_row',
+            'name': name,
+            'row_id': find_row_by_name(name)['row_id'],
+            'why': why,
+            **(more_data or {})
+        })
+        names_only_ss.remove(ss_name)
+
+    # First, if a repo is only in the spreadsheet, check if it was
+    # moved or archived
+    for ss_name in names_only_ss.copy():
+        try:
+            found_repo = hub.repository(*ss_name.split('/', 2))
```

Comment:
Is this too opaque? I can't tell.

Reply:
Are you passing in the org and repo name to hub?  I think a comment would be good.

Reply:
OK. I'll just name the parts and pass them in explicitly.